libchisel/src/remapstart.rs

```diff
@@ -0,0 +1,262 @@
+use parity_wasm::elements::*;
+use std::collections::HashMap;
+
+use super::{ModuleError, ModulePreset, ModuleTranslator};
+
+pub struct RemapStart;
+
+impl ModulePreset for RemapStart {
+    fn with_preset(preset: &str) -> Result<Self, ()> {
+        match preset {
+            // TODO refactor this later
+            "ewasm" => Ok(RemapStart {}),
+            _ => Err(()),
+        }
+    }
+}
+
+impl ModuleTranslator for RemapStart {
+    fn translate_inplace(&self, module: &mut Module) -> Result<bool, ModuleError> {
+        Ok(remap_start(module))
+    }
+
+    fn translate(&self, module: &Module) -> Result<Option<Module>, ModuleError> {
+        let mut ret = module.clone();
+        if remap_start(&mut ret) {
+            Ok(Some(ret))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+fn clear_names_section(module: &mut Module) {
+    let sections = module.sections_mut();
+    let mut rmidx = sections.len();
+    for (index, section) in sections.iter_mut().enumerate() {
+        if let Section::Name(_) = section {
+            rmidx = index;
+            break;
+        }
+    }
+    if rmidx < sections.len() {
+        sections.remove(rmidx);
+    }
+}
+
+// replace an exported function with another function
+fn replace_export(module: &mut Module, export_name: &str, func_idx: u32) -> bool {
+    let mut ret = false;
+
+    if let Some(mut export_section) = module.export_section_mut() {
+        for export_entry in export_section.entries_mut().iter_mut() {
+            if *export_entry.field() == export_name.to_string() {
+                *export_entry =
+                    ExportEntry::new(export_name.to_string(), Internal::Function(func_idx));
+                ret = true;
+            }
+        }
+    } else {
+        module
+            .sections_mut()
+            .push(Section::Export(ExportSection::with_entries(vec![
+                ExportEntry::new(export_name.to_string(), Internal::Function(func_idx)),
```

Comment:
Since all these functions need `export_name` to be a `String` I'd argue just change the function singature to use that or to have a `let export_name = export_name.to_string()` on the op.

Reply:
I think it is not an issue here because the pwasm functions consume strings, so converting a `&str` to owned form inline is cleaner than instantiating a bunch of clones upfront to move into the functions/structs.

Reply:
Bunch of clones? Single variable.

Reply:
Yes but if you move a string you pass ownership. it is necessary to make copies for pwasm to consume because it does not accept by reference